test/bugle_forms/handlers/user_test.clj

```diff
@@ -28,5 +28,41 @@
                               :password "t0ps3cr3t"}
           request {:form-params signup-form-params}
           response (sut/create-user request)]
-      (is (= 302 (:status response)))
+      (is (= 303 (:status response)))
       (is (= {"Location" "/signup"} (:headers response))))))
+
+(deftest login-and-logout
+  (let [signup-form-params {:name "Ghanshyam Pinto"
+                            :email "gs@nilen.so"
+                            :password "t0ps3cr3t"}
+        signup-request {:form-params signup-form-params}
+        _ (sut/create-user signup-request)]
+    (testing "Successful login has session attached and redirects to dashboard"
+      (let [login-form-params (select-keys signup-form-params
+                                           [:email :password])
+            login-request {:form-params login-form-params}
+            response (sut/login login-request)]
+        (is (:session response))
+        (is (= 303 (:status response)))
+        (is (= {"Location" "/dashboard"} (:headers response)))))
+
+    (testing "Unsuccessful login redirects to login page again"
+      (let [login-form-params {:email (:email signup-form-params)
+                               :password "hunter2"}
+            login-request {:form-params login-form-params}
+            response (sut/login login-request)]
+        (is (= 303 (:status response)))
+        (is (nil? (get-in response [:headers "Set-Cookie"])))
```

Comment:
Since the request is made directly to the user handler, the response here won't have any headers for setting the session cookie - it is added when it passes through the session middleware in the middleware chain.

At this point, the tests for session should be just based on `:session` in response map

Reply:
Got it, thanks for catching this!